src/Common/MemorySpillScheduler.cpp

```diff
@@ -0,0 +1,69 @@
+#include <mutex>
+#include <Common/MemoryTracker.h>
+#include <Common/MemoryTrackerUtils.h>
+#include <Common/MemorySpillScheduler.h>
+#include <Processors/IProcessor.h>
+
+#include <Poco/Logger.h>
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+void MemorySpillScheduler::checkAndSpill(IProcessor * processor)
+{
+
+    if (!enable || !processor->spillable() || !getHardLimit())
```

Comment:
We have already checked that `processor->spillable()` is true before calling this function, let's assume it is true here. Maybe just add a `chassert()`.